### ISSUE TITLE
Condition System (new version)

### ISSUE THREAD

**Issue by puf17640** [fluentscheduler/FluentScheduler#216]
### I am working on an "**If this, then that**" type of application which uses the current version of **FluentScheduler**. 
_I tweaked the library a little bit to fit my use but that's not what I want to discuss in here._

What I really missed in the current version is a **condition mechanism**, simply beeing able to check if some arbitrary condition is fullfilled before executing the task, eg. checking if a process is running before starting another. So without much thinking I went straight into hacking and it is working perfectly fine as of now, but since I haven't taken much time into thinking it out, it might not be the best way to implement it.
 
I have added a ICondition interface which defines the method Evaluate() and a ConditionFactory which is just the equivalent of JobFactory (but using IConditon of course)
```
public interface ICondition
{
    bool Evaluate();
}
```
Example Implementation ->
```
class IsBatteryAbove15Condition : ICondition
{
    public bool Evaluate() => SystemInformation.PowerStatus.BatteryLifePercent > .15f;
}
```
I also added methods to Schedule.cs and JobManager.cs
**Schedule.cs ->**
```
internal Func<bool> Condition { get; private set; }

public Schedule(IEnumerable<Action> actions, Func<bool> condition)
{
    Disabled = false;
    Condition = condition;
    Jobs = actions.ToList();
    AdditionalSchedules = new List<Schedule>();
    PendingRunOnce = false;
    Reentrant = null;
}

public Schedule WithCondition(ICondition condition)
{
    if (condition == null)
        throw new ArgumentNullException("condition");
    Condition = JobManager.GetConditionFunc(condition);
    return this;
}
```
**JobManager.cs ->**
```
internal static void RunJob(Schedule schedule)
{
    if (schedule.Disabled  || (schedule.Condition != null && !schedule.Condition.Invoke()))
        return;
    ...
}
```

I would really like to hear what you all think about this and discuss what could be done better! Implementing it in the new version of FluentScheduler would be awesome! I personally haven't thought about making this public in any way before because I didn't think anyone would need such feature in a task-scheduling library, but now it seems like that would be a nice to have feature. 

PS: Any ideas for my project? Anything is helpful! :)

-Julian

**Comment by puf17640** [comment]
Thanks for the answer, I just did it the way I did because it was for a school project and I had to have a set of Jobs and a set of Conditions which could be combined however you want to. I now realise it is actually not of value for anyone else, sorry.

**Comment by tallesl** [comment]
Don't be sorry. This sort of experimentation and stretching on what something can and should do is a great exercise, but not that helpful when it comes to shippable code.

Making a production-ready codebase is not all fun and games. There's a phrase that I often say to colleagues: "in an abstraction, more important than what you provide is what you don't".

I hope I didn't come as (overly) harsh.
